services/webtheme/webtheme.go

```diff
@@ -12,63 +13,146 @@ import (
 	"code.gitea.io/gitea/modules/log"
 	"code.gitea.io/gitea/modules/public"
 	"code.gitea.io/gitea/modules/setting"
+	"code.gitea.io/gitea/modules/util"
 )
 
 var (
-	availableThemes    []string
-	availableThemesSet container.Set[string]
-	themeOnce          sync.Once
+	availableThemes             []*ThemeMetaInfo
+	availableThemeInternalNames container.Set[string]
+	themeOnce                   sync.Once
 )
 
+const (
+	fileNamePrefix = "theme-"
+	fileNameSuffix = ".css"
+)
+
+type ThemeMetaInfo struct {
+	FileName     string
+	InternalName string
+	DisplayName  string
+}
+
+func parseThemeMetaInfoToMap(cssContent string) map[string]string {
+	/*
+		The theme meta info is stored in the CSS file's variables of `gitea-theme-meta-info` element,
+		which is a privately defined and is only used by backend to extract the meta info.
+		Not using ":root" because it is difficult to parse various ":root" blocks when importing other files,
+		it is difficult to control the overriding, and it's difficult to avoid user's customized overridden styles.
+	*/
+	metaInfoContent := cssContent
+	if pos := strings.LastIndex(metaInfoContent, "gitea-theme-meta-info"); pos >= 0 {
+		metaInfoContent = metaInfoContent[pos:]
+	}
+
+	reMetaInfoItem := `
+(
+\s*(--[-\w]+)
+\s*:
+\s*("(\\"|[^"])*")
+\s*;
+\s*
+)
+`
+	reMetaInfoItem = strings.ReplaceAll(reMetaInfoItem, "\n", "")
+	reMetaInfoBlock := `\bgitea-theme-meta-info\s*\{(` + reMetaInfoItem + `+)\}`
+	re := regexp.MustCompile(reMetaInfoBlock)
+	matchedMetaInfoBlock := re.FindAllStringSubmatch(metaInfoContent, -1)
+	if len(matchedMetaInfoBlock) == 0 {
+		return nil
+	}
+	re = regexp.MustCompile(strings.ReplaceAll(reMetaInfoItem, "\n", ""))
+	matchedItems := re.FindAllStringSubmatch(matchedMetaInfoBlock[0][1], -1)
+	m := map[string]string{}
+	for _, item := range matchedItems {
+		v := item[3]
+		v = strings.TrimPrefix(v, "\"")
+		v = strings.TrimSuffix(v, "\"")
+		v = strings.ReplaceAll(v, `\"`, `"`)
+		m[item[2]] = v
+	}
+	return m
+}
```

Comment:
Function needs a test. My version already had a test. Why did you remove it?

Reply:
My method doesn't need that test, it only uses the last `gitea-theme-meta-info` block.

If you'd like to test it, I will add one.

ps: I didn't remove it, as I said: I reset the branch to my method. Your work is saved in https://github.com/wxiaoguang/gitea/commits/theme-meta-info-silverwind/ . This is my original code without any further change.

Reply:
Added more tests and comment: fdfd4400615afc19d4d42524aeac1f8dffa6ed2d